sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala

```diff
@@ -0,0 +1,425 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector
+
+import java.util
+import java.util.Collections
+
+import test.org.apache.spark.sql.connector.catalog.functions.{JavaAverage, JavaStrLen}
+import test.org.apache.spark.sql.connector.catalog.functions.JavaStrLen._
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.{AnalysisException, Row}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.connector.catalog.{Identifier, SupportsNamespaces}
+import org.apache.spark.sql.connector.catalog.functions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+class DataSourceV2FunctionSuite extends DatasourceV2SQLBase {
+  private val emptyProps: util.Map[String, String] = Collections.emptyMap[String, String]
+
+  private def addFunction(ident: Identifier, fn: UnboundFunction): Unit = {
+    catalog("testcat").asInstanceOf[V2InMemoryCatalog].createFunction(ident, fn)
+  }
+
+  test("undefined function") {
+    assert(intercept[AnalysisException](
+      sql("SELECT testcat.non_exist('abc')").collect()
+    ).getMessage.contains("Undefined function"))
+  }
+
+  test("non-function catalog") {
+    spark.conf.set("spark.sql.catalog.testcat", classOf[BasicInMemoryTableCatalog].getName)
```

Comment:
Shouldn't these tests use `withConf` to ensure the configuration settings are removed?

Reply:
Good point! fixed.